File: bindings/java-jni/src/main/java/org/maplibre/nativejni/map/MapProjectionHandle.java
package org.maplibre.nativejni.map;

import java.util.List;
import java.util.Objects;
import org.maplibre.nativejni.camera.CameraOptions;
import org.maplibre.nativejni.camera.EdgeInsets;
import org.maplibre.nativejni.geo.Geometry;
import org.maplibre.nativejni.geo.LatLng;
import org.maplibre.nativejni.geo.ScreenPoint;
import org.maplibre.nativejni.internal.access.InternalAccess;
import org.maplibre.nativejni.internal.javacpp.JavaCppSupport;
import org.maplibre.nativejni.internal.javacpp.JavaCppValues;
import org.maplibre.nativejni.internal.javacpp.MaplibreNativeC;
import org.maplibre.nativejni.internal.lifecycle.HandleState;
import org.maplibre.nativejni.internal.loader.NativeLibrary;
import org.maplibre.nativejni.internal.status.Status;
import org.maplibre.nativejni.internal.struct.CoreStructs;
import org.maplibre.nativejni.internal.struct.MapStructs;

/** Owned standalone projection snapshot created from a map. */
public final class MapProjectionHandle implements AutoCloseable {
  private final HandleState state;

  private MapProjectionHandle(long handle) {
    this.state = new HandleState("MapProjectionHandle", handle);
  }

  public static MapProjectionHandle create(MapHandle map) {
    Objects.requireNonNull(map, "map");
    NativeLibrary.ensureLoaded();
    var outProjection = JavaCppSupport.outPointer(MaplibreNativeC.mln_map_projection.class);
    Status.check(
        MaplibreNativeC.mln_map_projection_create(
            JavaCppSupport.map(map.nativeAddress(InternalAccess.INSTANCE)), outProjection));
    return new MapProjectionHandle(
        JavaCppSupport.outAddress(outProjection, MaplibreNativeC.mln_map_projection.class));
  }

  public CameraOptions camera() {
    NativeLibrary.ensureLoaded();
    var outCamera = MaplibreNativeC.mln_camera_options_default();
    Status.check(
        MaplibreNativeC.mln_map_projection_get_camera(
            JavaCppSupport.projection(state.requireLiveAddress()), outCamera));
    return MapStructs.cameraOptions(outCamera);
  }

  public void setCamera(CameraOptions camera) {
    NativeLibrary.ensureLoaded();
    Objects.requireNonNull(camera, "camera");
    var nativeCamera = MapStructs.nativeCameraOptions(camera);
    try {
      Status.check(
          MaplibreNativeC.mln_map_projection_set_camera(
              JavaCppSupport.projection(state.requireLiveAddress()), nativeCamera));
    } finally {
      nativeCamera.close();
    }
  }

  public void setVisibleCoordinates(List<LatLng> coordinates, EdgeInsets padding) {
    NativeLibrary.ensureLoaded();
    Objects.requireNonNull(coordinates, "coordinates");
    Objects.requireNonNull(padding, "padding");
    var copiedCoordinates = List.copyOf(coordinates);
    try (var nativeCoordinates = CoreStructs.latLngArray(copiedCoordinates);
        var nativePadding = CoreStructs.nativeEdgeInsets(padding)) {
      Status.check(
          MaplibreNativeC.mln_map_projection_set_visible_coordinates(
              JavaCppSupport.projection(state.requireLiveAddress()),
              nativeCoordinates.coordinates(),
              nativeCoordinates.count(),
              nativePadding));
    }
  }

  public void setVisibleGeometry(Geometry geometry, EdgeInsets padding) {
    NativeLibrary.ensureLoaded();
    Objects.requireNonNull(geometry, "geometry");
    Objects.requireNonNull(padding, "padding");
    try (var nativeGeometry = JavaCppValues.geometry(geometry);
        var nativePadding = CoreStructs.nativeEdgeInsets(padding)) {
      Status.check(
          MaplibreNativeC.mln_map_projection_set_visible_geometry(
              JavaCppSupport.projection(state.requireLiveAddress()),
              nativeGeometry.value(),
              nativePadding));
    }
  }

  public ScreenPoint pixelForLatLng(LatLng coordinate) {
    NativeLibrary.ensureLoaded();
    Objects.requireNonNull(coordinate, "coordinate");
    try (var nativeCoordinate = CoreStructs.nativeLatLng(coordinate)) {
      var outPoint = new MaplibreNativeC.mln_screen_point();
      Status.check(
          MaplibreNativeC.mln_map_projection_pixel_for_lat_lng(
              JavaCppSupport.projection(state.requireLiveAddress()), nativeCoordinate, outPoint));
      return CoreStructs.screenPoint(outPoint);
    }
  }

  public LatLng latLngForPixel(ScreenPoint point) {
    NativeLibrary.ensureLoaded();
    Objects.requireNonNull(point, "point");
    try (var nativePoint = CoreStructs.nativeScreenPoint(point)) {
      var outCoordinate = new MaplibreNativeC.mln_lat_lng();
      Status.check(
          MaplibreNativeC.mln_map_projection_lat_lng_for_pixel(
              JavaCppSupport.projection(state.requireLiveAddress()), nativePoint, outCoordinate));
      return CoreStructs.latLng(outCoordinate);
    }
  }

  public void close() {
    state.closeOnce(
        address -> MaplibreNativeC.mln_map_projection_destroy(JavaCppSupport.projection(address)));
  }

  public boolean isClosed() {
    return state.isReleased();
  }
}

File: bindings/dotnet/src/Maplibre.Native/Map/MapProjectionHandle.cs
using Maplibre.Native.Camera;
using Maplibre.Native.Geo;
using Maplibre.Native.Internal.C;
using Maplibre.Native.Internal.Handle;
using Maplibre.Native.Internal.Status;
using Maplibre.Native.Internal.Struct;

namespace Maplibre.Native.Map;

/// <summary>Owner-thread projection snapshot handle.</summary>
public sealed unsafe class MapProjectionHandle : IDisposable
{
    private readonly NativeHandleState<mln_map_projection> state;

    private MapProjectionHandle(mln_map_projection* handle)
    {
        state = new NativeHandleState<mln_map_projection>(
            handle,
            static handle => NativeMethods.mln_map_projection_destroy(handle),
            nameof(MapProjectionHandle)
        );
    }

    internal static MapProjectionHandle Create(MapHandle map)
    {
        ArgumentNullException.ThrowIfNull(map);
        mln_map_projection* projection = null;
        NativeStatus.Check(NativeMethods.mln_map_projection_create(map.Pointer, &projection));
        return new MapProjectionHandle(projection);
    }

    internal mln_map_projection* Pointer => state.Pointer;

    /// <summary>Whether this wrapper has successfully closed its native handle.</summary>
    public bool IsClosed => state.IsClosed;

    /// <summary>Gets the projection camera descriptor.</summary>
    public CameraOptions GetCamera()
    {
        var camera = NativeMethods.mln_camera_options_default();
        NativeStatus.Check(NativeMethods.mln_map_projection_get_camera(Pointer, &camera));
        return MapStructs.CameraOptionsFromNative(camera);
    }

    /// <summary>Sets the projection camera descriptor, applying only non-null fields.</summary>
    public void SetCamera(CameraOptions camera)
    {
        var nativeCamera = MapStructs.ToNative(camera);
        NativeStatus.Check(NativeMethods.mln_map_projection_set_camera(Pointer, &nativeCamera));
    }

    /// <summary>Sets a camera that makes the supplied coordinates visible with padding.</summary>
    public void SetVisibleCoordinates(IReadOnlyList<LatLng> coordinates, EdgeInsets padding)
    {
        ArgumentNullException.ThrowIfNull(coordinates);
        var nativeCoordinates = new mln_lat_lng[coordinates.Count];
        for (var index = 0; index < coordinates.Count; index++)
        {
            nativeCoordinates[index] = CoreStructs.ToNative(coordinates[index]);
        }

        var nativePadding = MapStructs.ToNative(padding);
        fixed (mln_lat_lng* coordinatesPointer = nativeCoordinates)
        {
            NativeStatus.Check(
                NativeMethods.mln_map_projection_set_visible_coordinates(
                    Pointer,
                    nativeCoordinates.Length == 0 ? null : coordinatesPointer,
                    (nuint)nativeCoordinates.Length,
                    nativePadding
                )
            );
        }
    }

    /// <summary>Sets a camera that makes the supplied geometry visible with padding.</summary>
    public void SetVisibleGeometry(Geometry geometry, EdgeInsets padding)
    {
        ArgumentNullException.ThrowIfNull(geometry);
        using var nativeGeometry = NativeGeometry.From(geometry);
        var nativePadding = MapStructs.ToNative(padding);
        NativeStatus.Check(
            NativeMethods.mln_map_projection_set_visible_geometry(
                Pointer,
                nativeGeometry.Pointer,
                nativePadding
            )
        );
    }

    /// <summary>Converts a geographic coordinate to a screen pixel using this projection snapshot.</summary>
    public ScreenPoint PixelForLatLng(LatLng coordinate)
    {
        var nativeCoordinate = CoreStructs.ToNative(coordinate);
        mln_screen_point point = default;
        NativeStatus.Check(
            NativeMethods.mln_map_projection_pixel_for_lat_lng(Pointer, nativeCoordinate, &point)
        );
        return MapStructs.FromNative(point);
    }

    /// <summary>Converts a screen pixel to a geographic coordinate using this projection snapshot.</summary>
    public LatLng LatLngForPixel(ScreenPoint point)
    {
        var nativePoint = MapStructs.ToNative(point);
        mln_lat_lng coordinate = default;
        NativeStatus.Check(
            NativeMethods.mln_map_projection_lat_lng_for_pixel(Pointer, nativePoint, &coordinate)
        );
        return CoreStructs.FromNative(coordinate);
    }

    /// <summary>Destroys the projection on its owner thread.</summary>
    public void Close()
    {
        state.Close();
    }

    /// <inheritdoc />
    public void Dispose()
    {
        state.TryClose();
    }
}
